// chordsheetjs/lib/chord_sheet/song.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _line = require('./line');

var _line2 = _interopRequireDefault(_line);

var _tag = require('./tag');

var _tag2 = _interopRequireDefault(_tag);

var _utilities = require('../utilities');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var TITLE = 'title';
var SUBTITLE = 'subtitle';

var Song = function () {
  function Song() {
    var metaData = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    _classCallCheck(this, Song);

    this.lines = [];
    this.currentLine = null;
    this.metaData = metaData;
  }

  _createClass(Song, [{
    key: 'chords',
    value: function chords(chr) {
      this.currentLine.chords(chr);
    }
  }, {
    key: 'lyrics',
    value: function lyrics(chr) {
      this.ensureLine();
      this.currentLine.lyrics(chr);
    }
  }, {
    key: 'addLine',
    value: function addLine() {
      this.currentLine = (0, _utilities.pushNew)(this.lines, _line2.default);
      return this.currentLine;
    }
  }, {
    key: 'addChordLyricsPair',
    value: function addChordLyricsPair() {
      this.ensureLine();
      return this.currentLine.addChordLyricsPair();
    }
  }, {
    key: 'dropLine',
    value: function dropLine() {
      this.lines.pop();
    }
  }, {
    key: 'ensureLine',
    value: function ensureLine() {
      if (!this.currentLine) {
        this.addLine();
      }
    }
  }, {
    key: 'addTag',
    value: function addTag(name, value) {
      var tag = new _tag2.default(name, value);

      if (tag.isMetaTag()) {
        this.metaData[tag.name] = tag.value;
      }

      this.ensureLine();
      this.currentLine.addTag(tag);

      return tag;
    }
  }, {
    key: 'title',
    get: function get() {
      return this.metaData[TITLE] || "";
    }
  }, {
    key: 'subtitle',
    get: function get() {
      return this.metaData[SUBTITLE] || "";
    }
  }]);

  return Song;
}();

exports.default = Song;
